rust_src/src/crypto/mod.rs

```diff
@@ -1,78 +1,470 @@
+) -> LispObject {
+    let mut string = object.as_string();
+    let buffer = object.as_buffer();
+    let input = get_input(
+        object,
+        &mut string,
+        &buffer,
+        start,
+        end,
+        coding_system,
+        noerror,
+    );
+    _secure_hash(HashAlg::HashMD5, input.as_slice(), true)
+}
+
+/// Return the secure hash of OBJECT, a buffer or string.
+/// ALGORITHM is a symbol specifying the hash to use:
+/// md5, sha1, sha224, sha256, sha384 or sha512.
+///
+/// The two optional arguments START and END are positions specifying for
+/// which part of OBJECT to compute the hash.  If nil or omitted, uses the
+/// whole OBJECT.
+///
+/// If BINARY is non-nil, returns a string in binary form.
+#[lisp_fn(min = "2")]
+fn secure_hash(
+    algorithm: LispObject,
+    object: LispObject,
+    start: LispObject,
+    end: LispObject,
+    binary: LispObject,
+) -> LispObject {
+    let mut string = object.as_string();
+    let buffer = object.as_buffer();
+    let input = get_input(
+        object,
+        &mut string,
+        &buffer,
+        start,
+        end,
+        LispObject::constant_nil(),
+        LispObject::constant_nil(),
+    );
+    _secure_hash(hash_alg(algorithm), input.as_slice(), binary.is_nil())
+}
 
 #[no_mangle]
-pub unsafe extern "C" fn sha1_buffer(
-    buffer: *const c_char,
-    len: size_t,
-    dest_buf: *mut c_void,
-) -> *mut c_void {
+fn _secure_hash(algorithm: HashAlg, input: &[u8], hex: bool) -> LispObject {
+    let digest_size: usize;
+    let hash_func: unsafe fn(&[u8], &mut [u8]);
+    match algorithm {
+        HashAlg::HashMD5 => {
+            digest_size = MD5_DIGEST_LEN;
+            hash_func = md5_buffer;
+        }
+        HashAlg::HashSHA1 => {
+            digest_size = SHA1_DIGEST_LEN;
+            hash_func = sha1_buffer;
+        }
+        HashAlg::HashSHA224 => {
+            digest_size = SHA224_DIGEST_LEN;
+            hash_func = sha224_buffer;
+        }
+        HashAlg::HashSHA256 => {
+            digest_size = SHA256_DIGEST_LEN;
+            hash_func = sha256_buffer;
+        }
+        HashAlg::HashSHA384 => {
+            digest_size = SHA384_DIGEST_LEN;
+            hash_func = sha384_buffer;
+        }
+        HashAlg::HashSHA512 => {
+            digest_size = SHA512_DIGEST_LEN;
+            hash_func = sha512_buffer;
+        }
+    }
+
+    let buffer_size = if hex {
+        (digest_size * 2) as EmacsInt
+    } else {
+        digest_size as EmacsInt
+    };
+    let digest = LispObject::from_raw(unsafe { make_uninit_string(buffer_size as i64) });
+    let digest_str = digest.as_string_or_error();
+    unsafe {
+        // we can call this safely because we know that we made
+        // digest's buffer long enough
+        hash_func(input, digest_str.as_mut_slice());
+    }
+    if hex {
+        make_digest_string(digest_str.as_mut_slice(), digest_size);
+    }
+    digest
+}
+
+fn make_digest_string(buffer: &mut [u8], len: usize) {
```

Comment:
This one is curious; len is the length of the digest (16 bytes for md5, for example), and the buffer is twice that so that each character of the digest can be expanded to 2 hex digits. Perhaps a better name is in order, or a comment.